pkg/alertmanager/multitenant_test.go

```diff
@@ -368,3 +395,456 @@ receivers:
+				tt.updateRing(ringDesc)
+				return ringDesc, true, nil
+			}))
+
+			// Assert if we expected a sync or not.
+			if tt.expected {
+				test.Poll(t, time.Second, float64(2), func() interface{} {
+					metrics := regs.BuildMetricFamiliesPerUser()
+					return metrics.GetSumOfCounters("cortex_alertmanager_sync_configs_total")
+				})
+			} else {
+				time.Sleep(250 * time.Millisecond)
+
+				metrics := regs.BuildMetricFamiliesPerUser()
+				assert.Equal(t, float64(1), metrics.GetSumOfCounters("cortex_alertmanager_sync_configs_total"))
+			}
+		})
+	}
+}
+
+func TestAlertmanager_RingLifecyclerShouldAutoForgetUnhealthyInstances(t *testing.T) {
+	const unhealthyInstanceID = "alertmanager-bad-1"
+	const heartbeatTimeout = time.Minute
+	ctx := context.Background()
+	amConfig := mockAlertmanagerConfig(t)
+	amConfig.ShardingEnabled = true
+	amConfig.ShardingRing.HeartbeatPeriod = 100 * time.Millisecond
+	amConfig.ShardingRing.HeartbeatTimeout = heartbeatTimeout
+
+	ringStore := consul.NewInMemoryClient(ring.GetCodec())
+	mockStore := &mockAlertStore{
+		configs: map[string]alerts.AlertConfigDesc{},
+	}
+
+	am, err := createMultitenantAlertmanager(amConfig, nil, nil, mockStore, ringStore, log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, am))
+	defer services.StopAndAwaitTerminated(ctx, am) //nolint:errcheck
+
+	require.NoError(t, ringStore.CAS(ctx, RingKey, func(in interface{}) (interface{}, bool, error) {
+		ringDesc := ring.GetOrCreateRingDesc(in)
+		instance := ringDesc.AddIngester(unhealthyInstanceID, "127.0.0.1", "", generateSortedTokens(RingNumTokens), ring.ACTIVE, time.Now())
+		instance.Timestamp = time.Now().Add(-(ringAutoForgetUnhealthyPeriods + 1) * heartbeatTimeout).Unix()
+		ringDesc.Ingesters[unhealthyInstanceID] = instance
+
+		return ringDesc, true, nil
+	}))
+
+	test.Poll(t, time.Second, false, func() interface{} {
+		d, err := ringStore.Get(ctx, RingKey)
+		if err != nil {
+			return err
+		}
+
+		_, ok := ring.GetOrCreateRingDesc(d).Ingesters[unhealthyInstanceID]
+		return ok
+	})
+}
+
+func TestAlertmanager_InitialSyncFailureWithSharding(t *testing.T) {
+	ctx := context.Background()
+	amConfig := mockAlertmanagerConfig(t)
+	amConfig.ShardingEnabled = true
+	ringStore := consul.NewInMemoryClient(ring.GetCodec())
+	mockStore := &mockAlertStore{
+		configs:     map[string]alerts.AlertConfigDesc{},
+		WithListErr: fmt.Errorf("a fetch list failure"),
+	}
+
+	am, err := createMultitenantAlertmanager(amConfig, nil, nil, mockStore, ringStore, log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	defer services.StopAndAwaitTerminated(ctx, am) //nolint:errcheck
+
+	require.NoError(t, am.StartAsync(ctx))
+	err = am.AwaitRunning(ctx)
+	require.Error(t, err)
+	require.Equal(t, services.Failed, am.State())
+	require.False(t, am.ringLifecycler.IsRegistered())
+	require.NotNil(t, am.ring)
+}
+
+func generateSortedTokens(numTokens int) ring.Tokens {
```

Comment:
You can directly use `GenerateTokens()` because now it returns sorted tokens (we changed it recently).